### PR TITLE
ALSA/HDA: fix hdac_stream and hdac_ext_stream locking

### PR DESCRIPTION
**Pull request by plbossart**
While looking at Hdaudio issues, I noticed the locking is rather questionable. Here's a tentative fix to use locking in more consistent ways.

### CODE REVIEW

File: sound/hda/ext/hdac_ext_stream.c [hunk 1]
```diff
@@ -378,15 +386,17 @@ void snd_hdac_ext_stream_release(struct hdac_ext_stream *stream, int type)
...
-			snd_hdac_ext_stream_decouple(bus, stream, false);
+			snd_hdac_ext_stream_decouple_locked(bus, stream, false);
+		spin_unlock_irq(&bus->reg_lock);
 		snd_hdac_stream_release(&stream->hstream);
 		break;
 
 	case HDAC_EXT_STREAM_TYPE_LINK:
+		spin_lock_irq(&bus->reg_lock);
 		if (stream->decoupled && !stream->hstream.opened)
 			snd_hdac_ext_stream_decouple(bus, stream, false);
```
Comment (RanderWang) [inline sound/hda/ext/hdac_ext_stream.c]: snd_hdac_ext_stream_decouple_locked ?

**Review by ujfalusi — CHANGES_REQUESTED** [review]
Looks good, but @RanderWang's comment is valid.

File: sound/hda/ext/hdac_ext_stream.c (hunk 1 above)
Comment (plbossart) [reply sound/hda/ext/hdac_ext_stream.c]: yes indeed, thanks for spotting this @RanderWang

Comment (plbossart) [reply sound/hda/ext/hdac_ext_stream.c]: fixed.

**Review by ujfalusi — DISMISSED** [review]

**Review by kv2019i — COMMENTED** [review]
The non-locking patches seem good. The locking changes are hard to review, because it's not really clear what these locks are really protecting. E.g. assign and release seem to be mostly used from pcm_open/close which is non-atomic context and with the PCM mutex held. So is there any need to use spin locks for these ops to begin with? 

Then there's a lot of reg_lock usage that is obviously needed, so this is really hard to review.

**Comment by plbossart** [comment]
I agree @kv2019i I initially thought that the spin_lock was only there to protect register read/write on the HDaudio controller. But this is the code that Takashi wrote initially, so I imagine he had a reason to do so. I can also send this as RFC to get confirmation that this locking is actually needed.

**Comment by kv2019i** [comment]
@plbossart  wrote:
> I agree @kv2019i I initially thought that the spin_lock was only there to protect register read/write on the HDaudio controller. > But this is the code that Takashi wrote initially, so I imagine he had a reason to do so. I can also send this as RFC to get 
> confirmation that this locking is actually needed.

It's really quite hard to parse. The reg_lock code goes back a long, long way. You can see most of the related hda code is moved from even older ac97 drivers, and at some point Takashi combined multiple distinct spinlocks under "reg_lock" (at that point the name lost its meaning). But yeah, there seems to be genuine needs for the lock as well, so looks a bit tedious to go through.

**Comment by plbossart** [comment]
Sent as RFC to alsa-devel mailing list.

File: sound/hda/hdac_stream.c [hunk 2]
```diff
@@ -309,13 +310,12 @@ struct hdac_stream *snd_hdac_stream_assign(struct hdac_bus *bus,
...
 	}
 	if (res) {
-		spin_lock_irq(&bus->reg_lock);
 		res->opened = 1;
 		res->running = 0;
 		res->assigned_key = key;
 		res->substream = substream;
-		spin_unlock_irq(&bus->reg_lock);
 	}
+	spin_unlock_irq(&bus->reg_lock);
```
Comment (lyakh) [inline sound/hda/hdac_stream.c:318]: Looking at the code, the only field that affects this code, that is changed in `snd_hdac_stream_release()` is `.opened`, I also assume, that `snd_hdac_stream_release()` can only be called when `.opened == 1`, it wouldn't be called on an unassigned stream, right? So, if `snd_hdac_stream_release()` is called before it is checked in line 302/303, `.opened` will be 0 and it can be used by `snd_hdac_stream_assign()`. If `snd_hdac_stream_release()` is called after the check, `.opened` is still 1 and the stream is skipped, which is also correct. The only potential problem, that I can see, is if `snd_hdac_stream_release()` and `snd_hdac_stream_assign()` run on different cores, then `snd_hdac_stream_release()` sets `.opened` to 0 before it is checked in line 302/303 above. Then even if `snd_hdac_stream_assign()` is running on a faster core, so that it continues to line 312/- while `snd_hdac_stream_release()` is still between lines 334 and 335 below, the spin-lock will force `snd_hdac_stream_assign()` to wait for `snd_hdac_stream_release()` to complete, so `snd_hdac_stream_assign()` will anyway produce a correct result. So, I don't see a problem there either. Could you describe the race that this patch is fixing?

File: sound/hda/hdac_stream.c [hunk 3]
```diff
@@ -333,6 +333,7 @@ void snd_hdac_stream_release(struct hdac_stream *azx_dev)
 	spin_lock_irq(&bus->reg_lock);
 	azx_dev->opened = 0;
 	azx_dev->running = 0;
+	azx_dev->assigned_key = 0;
```
Comment (lyakh) [inline sound/hda/hdac_stream.c:336]: Confused. `snd_hdac_stream_assign()` is looking for closed streams with the same key:

```
		if (azx_dev->opened)
			continue;
		if (azx_dev->assigned_key == key) {
			res = azx_dev;
			break;
		}
```

After this it seems like it will never find such streams, because once a stream is released its `.assigned_key` is set to an invalid value?

File: sound/hda/ext/hdac_ext_stream.c [hunk 4]
```diff
@@ -106,20 +106,14 @@ void snd_hdac_stream_free_all(struct hdac_bus *bus)
...
 EXPORT_SYMBOL_GPL(snd_hdac_stream_free_all);
 
-/**
- * snd_hdac_ext_stream_decouple - decouple the hdac stream
- * @bus: HD-audio core bus
- * @stream: HD-audio ext core stream object to initialize
- * @decouple: flag to decouple
- */
-void snd_hdac_ext_stream_decouple(struct hdac_bus *bus,
-				struct hdac_ext_stream *stream, bool decouple)
```
Comment (lyakh) [inline sound/hda/ext/hdac_ext_stream.c:116]: do I understand it correctly, that this function uses a double-negation logic? It's called `decouple` (not-couple) and has a `decouple` (not-couple) parameter, so, calling `decouple(false)` means not-not-couple, i.e. do the coupling?.. Wouldn't it be better to split it into 2 functions: `couple()` and `decouple()` with no flags? There's practically 0 common code anyway. Then verifying locking should also become easier...

File: sound/hda/ext/hdac_ext_stream.c [hunk 5]
```diff
@@ -378,15 +386,17 @@ void snd_hdac_ext_stream_release(struct hdac_ext_stream *stream, int type)
...
 		if (stream->decoupled && !stream->link_locked)
-			snd_hdac_ext_stream_decouple(bus, stream, false);
+			snd_hdac_ext_stream_decouple_locked(bus, stream, false);
+		spin_unlock_irq(&bus->reg_lock);
 		snd_hdac_stream_release(&stream->hstream);
 		break;
 
 	case HDAC_EXT_STREAM_TYPE_LINK:
-		if (stream->decoupled && !stream->hstream.opened)
-			snd_hdac_ext_stream_decouple(bus, stream, false);
```
Comment (lyakh) [inline sound/hda/ext/hdac_ext_stream.c:388]: looks like `snd_hdac_ext_stream_release()` can be called on a closed stream?

**Review by lyakh — COMMENTED** [review]
Not very clear to me. In general this code is called from hw_params / hw_free / open flows, right? Since it uses single boolean flags instead of use-counts, it looks like these interfaces are supposed to be only opened once. Do we observe any races in this code? Would it be possible and wouldn't it be better to protect those flows with mutexes at a higher level, provided they're never called from atomic contexts?

**Comment by plbossart** [comment]
> Not very clear to me. In general this code is called from hw_params / hw_free / open flows, right? Since it uses single boolean flags instead of use-counts, it looks like these interfaces are supposed to be only opened once. Do we observe any races in this code? Would it be possible and wouldn't it be better to protect those flows with mutexes at a higher level, provided they're never called from atomic contexts?

No some of these functions are called in TRIGGER_STOP or TRIGGER_PAUSE_PUSH.
I don't have a very good understanding of what is supposed to be protected, that's why I shared these patches on alsa-devel as RFC. The point is that the spin_lock used *today* makes no sense.

**Comment by kv2019i** [comment]
@plbossart This can be closed, right (already merged usptream)?

**Comment by plbossart** [comment]
> @plbossart This can be closed, right (already merged usptream)?

no, it was sent as RFC and there are *zero* comments so far...

**Comment by kv2019i** [comment]
@plbossart wrote: 
> > @plbossart This can be closed, right (already merged usptream)?
> 
> no, it was sent as RFC and there are _zero_ comments so far...

Are we on the same list? :)

20f3b10de61add5e14b6ce4df982f4df2a4cbbc (tiwai/for-next) ASoC: SOF: Intel: hda-dai: fix potential locking issue
868ddfcef31ff93ea8961b2e81ea7fe12f6f144b ALSA: hda: hdac_ext_stream: fix potential locking issues
1465d06a6d8580e73ae65f8590392df58c5ed2fd ALSA: hda: hdac_stream: fix potential locking issue in snd_hdac_stream_assign()

**Comment by plbossart** [comment]
> @plbossart wrote:
> 
> > > @plbossart This can be closed, right (already merged usptream)?
> > 
> > 
> > no, it was sent as RFC and there are _zero_ comments so far...
> 
> Are we on the same list? :)
> 
> 20f3b10de61add5e14b6ce4df982f4df2a4cbbc (tiwai/for-next) ASoC: SOF: Intel: hda-dai: fix potential locking issue [868ddfc](https://github.com/thesofproject/linux/commit/868ddfcef31ff93ea8961b2e81ea7fe12f6f144b) ALSA: hda: hdac_ext_stream: fix potential locking issues [1465d06](https://github.com/thesofproject/linux/commit/1465d06a6d8580e73ae65f8590392df58c5ed2fd) ALSA: hda: hdac_stream: fix potential locking issue in snd_hdac_stream_assign()

ok, so 3 of the 4 patches were added but there was no 'applied, thanks' response, and feedback was indeed provided on another path. my email client didn't see them. thanks for correcting me @kv2019i
